loki/transformations/hoist_variables.py

```diff
@@ -319,6 +330,27 @@ def driver_variable_declaration(self, routine, variables):
         """
         routine.variables += tuple(v.rescope(routine) for v in variables)
 
+    @staticmethod
+    def driver_variable_declaration_dim_remapping(routine, variables):
+        """
+        Take a list of variables and remap their dimensions for those being
+        arrays to account for possibly uninitialized variables/dimensions. 
+
+        Parameters
+        ----------
+        routine : :any:`Subroutine`
+            The relevant subroutine.
+        variables : tuple of :any:`Variable`
+            The tuple of variables for remapping.
+        """
+        dim_map = {}
+        assignments = FindNodes(Assignment).visit(routine.body)
+        for assignment in assignments:
+            dim_map[assignment.lhs] = assignment.rhs
```

Comment:
Just to point out how to do this in a "single line", or single list comprehension at least. For larger kernels like CLOUDSC with hundreds of assignments I expect you might actually be able to measure a small advantage with below change.
```suggestion
        dim_map = {
            assignment.lhs: assignment.rhs
            for assignment in FindNodes(Assignment).visit(routine.body)
        }
```

Also, this map might become quite big, so we might want to build a list of candidates first and then create the mapping? Suggestion (untested) would be something like this:
```suggestion
        dim_vars = [
            dim_var
            for var in variables
            for dim_var in FindVariables().visit(var.dimensions)
            if isinstance(var, sym.Array)
        ]
        dim_map = {
            assignment.lhs: assignment.rhs
            for assignment in FindNodes(Assignment).visit(routine.body)
            if assignment.lhs in dim_vars
        }
```